client/boot/common.js

```diff
@@ -427,13 +438,42 @@ const boot = ( currentUser, registerRoutes ) => {
 			registerRoutes();
 		}
 
+		const isDesktop = config.isEnabled( 'desktop' );
+		const loggedIn = currentUser.get() !== false;
+
+		const render = () => {
+			renderLayout( reduxStore );
+			page.start( { decodeURLComponents: false } );
+		};
+
+		const renderDesktopPromise = () => {
+			return new Promise( function ( resolve ) {
+				const ipc = require( 'electron' ).ipcRenderer;
+
+				ipc.on( 'cookie-auth-complete', function () {
+					debug( 'Desktop cookies set, rendering main layout...' );
+					render();
+					resolve();
+				} );
```

Comment:
The best place for waiting for `cookie-auth-complete` is the `bootApp` function at the end of this file. The code would be:
```js
await user.initialize();
if ( config.isEnabled( 'desktop' ) ) {
  await waitForCookieAuth();
}
boot( user, registerRoutes );
```
This way, we wait for the user object to be available and for the cookies to be set before continuing with the boot sequence.

Reply:
That would be a much more compact way to do it, except this approach would be a non-trivial refactor given the `boot` function is what triggers the cookie authentication logic in the Main process. In sequence:

1. `bootApp` calls `boot`
2. `boot` calls `setupMiddlewares` ([link](https://github.com/Automattic/wp-calypso/blob/d682d34136b055b7c392ab3586f94c77de6a6411/client/boot/common.js#L424))
3. `setupMiddlewares` calls `require( 'lib/desktop' ).default.init()` ([link](https://github.com/Automattic/wp-calypso/blob/d682d34136b055b7c392ab3586f94c77de6a6411/client/boot/common.js#L383-L385))
4. Within `require( 'lib/desktop' ).default.init()`, there is a call to `sendUserLoginStatus`, which sends the IPC `user-auth` event ([link](https://github.com/Automattic/wp-calypso/blob/d682d34136b055b7c392ab3586f94c77de6a6411/client/lib/desktop/index.js#L83)).
5. The main process responds to `user-auth` IPC event by fetching and setting cookies to the Window's webContents session ([link](https://github.com/Automattic/wp-calypso/blob/d682d34136b055b7c392ab3586f94c77de6a6411/client/desktop/lib/cookie-auth/index.js#L101)).

The TL;DR here is that the trigger for cookie authentication happens within the `boot` function, making this approach really tricky.

I actually attempted this refactor in 1d21ee449e4e4f5af87cb508baa5c084ff78d15e, and the refactor was non-trivial. I ended up reverting as it wasn't quite working and failed with a timeout.

Happy to revisit this for a refactor later, but in the mean time we're shooting to get this patch into our upcoming release scheduled for Monday.

Reply:
> Happy to enhance/tweak this refactor later, but in the mean time we're shooting to get this patch into our upcoming release scheduled for Monday.

That's fair -- I'm two weeks late with my review, so I can't expect you to wait for me being 100% happy with the PR before shipping it 🙂 Let's improve both the Pinghub and this PR's code in follow-ups. I believe there's plenty of accidental complexity that can be removed.